File: src/main/java/seclogin/Password.java
package seclogin;

import java.util.Arrays;

import static com.google.common.base.Preconditions.checkNotNull;

/** A regular, unhardened, character sequence password. */
public class Password {

    private final char[] password;

    public Password(String password) {
        this.password = checkNotNull(password).toCharArray();
    }

    public char[] asCharArray() {
        return password;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Password that = (Password) o;

        if (!Arrays.equals(password, that.password)) return false;

        return true;
    }

    @Override
    public int hashCode() {
        return Arrays.hashCode(password);
    }
}

File: src/main/scala/seclogin/Password.scala
package seclogin

import java.nio.charset.Charset
import java.nio.CharBuffer

/** A user's unhardened password.
  */
class Password(string: String) {

  def asString: String = string

  def asChars: Array[Char] = string.toCharArray

  def asBytes: Array[Byte] =
    Charset.forName("UTF-8").encode(CharBuffer.wrap(string)).array

}

